src/vsphere_cpi/spec/integration/attach_tag_to_vm_spec.rb

```diff
@@ -0,0 +1,313 @@
+        begin
+          test_vm_id = @cpi.create_vm(
+            'agent-007',
+            @stemcell_id,
+            vm_type,
+            network_spec,
+            [],
+            {}
+          )
+          mob_id = @cpi.vm_provider.find(test_vm_id).mob.__mo_id__
+          expect(test_vm_id).to_not be_nil
+          expect(verify_tags(mob_id, ['tag-name-a-1', 'tag-name-a-2', 'tag-name-a-3', 'tag-name-c-1'])).to be(true)
+        ensure
+          delete_vm(@cpi, test_vm_id)
+        end
+      end
+    end
+
+    context 'when create a vm and attach a vm-non-associable tag to it' do
+      let(:tags) { [ { 'category' => 'category-name-b', 'tag' => 'tag-name-b-1'} ] }
+      it 'created a vm without tags attached' do
+        begin
+          test_vm_id = @cpi.create_vm(
+            'agent-007',
+            @stemcell_id,
+            vm_type,
+            network_spec,
+            [],
+            {}
+          )
+          mob_id = @cpi.vm_provider.find(test_vm_id).mob.__mo_id__
+          expect(test_vm_id).to_not be_nil
+          expect(verify_tags(mob_id, [])).to be(true)
+        ensure
+          delete_vm(@cpi, test_vm_id)
+        end
+      end
+    end
+
+    context 'when create a vm and attach multiple same-category-single-cardinality tags to it' do
+      let(:tags) {
+        [
+          { 'category' => 'category-name-a', 'tag' => 'tag-name-a-1'},
+          { 'category' => 'category-name-c', 'tag' => 'tag-name-c-1'},
+          { 'category' => 'category-name-c', 'tag' => 'tag-name-c-2'}
+        ]
+      }
+      it 'created a vm without same-category-single-cardinality tags' do
+        begin
+          test_vm_id = @cpi.create_vm(
+            'agent-007',
+            @stemcell_id,
+            vm_type,
+            network_spec,
+            [],
+            {}
+          )
+          mob_id = @cpi.vm_provider.find(test_vm_id).mob.__mo_id__
+          expect(test_vm_id).to_not be_nil
+          expect(verify_tags(mob_id, ['tag-name-a-1'])).to be(true)
+        ensure
+          delete_vm(@cpi, test_vm_id)
+        end
+      end
+    end
+
+    context 'when create a vm and attach wrong category tags to it' do
+      let(:tags) {
+        [
+          { 'category' => 'category-name-a', 'tag' => 'tag-name-a-1'},
+          { 'category' => 'category-name-d', 'tag' => 'tag-name-d-1'},
+        ]
+      }
+      it 'created a vm with correct category tags attached' do
+        begin
+          test_vm_id = @cpi.create_vm(
+            'agent-007',
+            @stemcell_id,
+            vm_type,
+            network_spec,
+            [],
+            {}
+          )
+          mob_id = @cpi.vm_provider.find(test_vm_id).mob.__mo_id__
+          expect(test_vm_id).to_not be_nil
+          expect(verify_tags(mob_id, ['tag-name-a-1'])).to be(true)
+        ensure
+          delete_vm(@cpi, test_vm_id)
+        end
+      end
+    end
+
+    context 'when create a vm and attach wrong tags to it' do
```

Comment:
change it to `when creating a VM and attach a non-existent category to it`

Reply:
Done.